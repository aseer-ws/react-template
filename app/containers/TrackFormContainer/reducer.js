import produce from 'immer';
import { createActions } from 'reduxsauce';

export const GENRE_ENUM = ['Jaz', 'HipHop', 'Electronics', 'Pop'];

export const initialValues = {
  trackName: '',
  trackPrice: 1,
  primaryGenreName: GENRE_ENUM[0],
  releaseDate: '22/01/2015',
  trackViewUrl: '',
  collectionName: '',
  collectionPrice: 2,
  collectionViewUrl: '',
  artistName: '',
  artistViewUrl: ''
};

export const fillValues = {
  trackName: 'Track A',
  trackPrice: 2,
  primaryGenreName: GENRE_ENUM[0],
  releaseDate: '02/11/2018',
  trackViewUrl: 'https://tracka.com',
  collectionName: 'Collection B',
  collectionPrice: 12,
  collectionViewUrl: 'https://collectionb.com',
  artistName: 'Artist C',
  artistViewUrl: 'https://artistc.com'
};

export const initialState = {
  formValues: initialValues
};

export const { Types: trackFormContainerTypes, Creators: trackFormContainerCreators } = createActions({
  setFormValues: ['formValues'],
  resetForm: {},
  fillForm: {}
});

const trackFormContainerReducer = function (state = initialState, action) {
  return produce(state, (draft) => {
    switch (action.type) {
      case trackFormContainerTypes.SET_FORM_VALUES:
        draft.formValues = { ...state.formValues, ...action.formValues };
        break;
      case trackFormContainerTypes.RESET_FORM:
        draft.formValues = initialValues;
        break;
      case trackFormContainerTypes.FILL_FORM:
        draft.formValues = fillValues;
        break;
      default:
        return initialState;
    }
  });
};

export default trackFormContainerReducer;
