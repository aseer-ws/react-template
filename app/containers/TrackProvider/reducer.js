/*
 *
 * TrackProvider reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = {
  list: { artistName: '', tracks: [], trackCount: 0, error: null },
  details: {
    trackId: null,
    track: {},
    error: null
  }
};

export const { Types: trackProviderTypes, Creators: trackProviderCreators } = createActions({
  requestGetTracks: ['artistName'],
  successGetTracks: ['data'],
  failureGetTracks: ['error'],
  clearTracks: {},
  requestGetTrack: ['trackId'],
  successGetTrack: ['data'],
  failureGetTrack: ['error']
});

/* eslint-disable default-case, no-param-reassign */
export const trackProviderReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case trackProviderTypes.REQUEST_GET_TRACKS:
        draft.list.artistName = action.artistName;
        break;
      case trackProviderTypes.SUCCESS_GET_TRACKS:
        draft.list.tracks = action.data.results;
        draft.list.trackCount = action.data.resultCount;
        draft.list.error = null;
        break;
      case trackProviderTypes.FAILURE_GET_TRACKS:
        draft.list.tracks = [];
        draft.list.trackCount = 0;
        draft.list.error = action.error;
        break;
      case trackProviderTypes.CLEAR_TRACKS:
        draft.list.tracks = [];
        draft.list.trackCount = 0;
        draft.list.error = null;
        break;
      case trackProviderTypes.REQUEST_GET_TRACK:
        draft.details.trackId = action.trackId;
        break;
      case trackProviderTypes.SUCCESS_GET_TRACK:
        draft.details.track = action.data;
        draft.details.error = null;
        break;
      case trackProviderTypes.FAILURE_GET_TRACK:
        draft.details.track = {};
        draft.details.error = action.error;
        break;
      default:
        return state;
    }
  });

export default trackProviderReducer;
