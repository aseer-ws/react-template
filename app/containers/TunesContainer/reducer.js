/*
 *
 * TunesContainer reducer
 *
 */
import produce from 'immer';
// import { fromJS } from 'immutable';
import { createActions } from 'reduxsauce';

export const initialState = {
  artist: '',
  songsData: {},
  tunesError: null
};

export const { Types: tunesContainerTypes, Creators: tunesContainerCreators } = createActions({
  setArtist: ['artistName'],
  setSongsData: ['songsData'],
  setTunesError: ['error'],
  clearSongs: {}
});

/* eslint-disable default-case, no-param-reassign */
export const tunesContainerReducer = produce((state = initialState, action) => {
  switch (action.type) {
    case tunesContainerTypes.SET_ARTIST:
      state.artist = action.artistName;
      break;
    case tunesContainerTypes.SET_SONGS_DATA:
      state.tunesError = null;
      state.songsData = action.songsData;
      break;
    case tunesContainerTypes.SET_TUNES_ERROR:
      state.songsData = {};
      state.tunesError = action.error;
      break;
    case tunesContainerTypes.CLEAR_SONGS:
      state.songsData = {};
      state.artist = '';
      break;
    default:
      return state;
  }
});

export default tunesContainerReducer;
