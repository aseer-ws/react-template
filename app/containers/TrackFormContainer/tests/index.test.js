/**
 *
 * Tests for TrackFormContainer container
 *
 *
 */

import history from '@app/utils/history';
import { fireEvent, waitFor } from '@testing-library/react';
import { renderProvider } from '@utils/testUtils';
import { message, notification } from 'antd';
import React from 'react';
import {
  INVALID_TYPE_MESSAGE,
  ISLT2M_MESSAGE,
  mapDispatchToProps,
  TrackFormContainerTest as TrackFormContainer,
  trackFormSteps
} from '../index';
import { initialValues, trackFormContainerCreators } from '../reducer';

const expectMessage = (baseElement, status, msg) =>
  expect(getMessageElm(baseElement, status)?.children[1].textContent).toBe(msg);

const getMessageElm = (baseElement, status) => baseElement.querySelector(`.ant-message-${status}`);
const getNotifElm = (baseElement) => baseElement.querySelector('.ant-notification-notice-content');

describe('<TrackFormContainer /> container tests', () => {
  let fillSpy;
  let resetSpy;
  let setFormSpy;
  let routeProps;

  beforeAll(() => {
    window.HTMLCanvasElement.prototype.getContext = function () {};
    window.URL.createObjectURL = function () {};
    history.location.pathname = '/add-track/track';
    fillSpy = jest.fn();
    resetSpy = jest.fn();
    setFormSpy = jest.fn();

    routeProps = {
      history,
      path: '/add-track/:stepTitle?'
    };
  });

  afterEach(async () => {
    fillSpy?.mockReset();
    resetSpy?.mockReset();
    setFormSpy?.mockReset();
    message.destroy();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should redirect first step if stepTitle params id is not provided', async () => {
    history.location.pathname = '/add-track';
    renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
  });

  it('should not move the next step if clicked on next step button without filling in data', async () => {
    const { getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
    const nextbtn = getByTestId('submit-btn');
    fireEvent.click(nextbtn);
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
  });

  it('should move to the next step if clicked on next step button with valid form data', async () => {
    const { getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));

    // validate current step fields to success
    fireEvent.change(document.getElementById('trackName'), { target: { value: 'Track A' } });
    fireEvent.change(document.getElementById('trackViewUrl'), { target: { value: 'https://tracka.com' } });

    const nextbtn = getByTestId('submit-btn');
    fireEvent.click(nextbtn);
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/collection'));
  });

  it('should fill the form and move to step 4 if clicked fill btn', async () => {
    const { getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    const fillBtn = getByTestId('fill-btn');
    fireEvent.click(fillBtn);
    const formItemToBeDisplayed = getByTestId(`${trackFormSteps[3][0].itemProps.name}_test`);
    expect(formItemToBeDisplayed).toBeInTheDocument();
  });

  it('should move to previous step if clicked on previous button', async () => {
    history.location.pathname = '/add-track/collection';
    const { getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    const prevBtn = getByTestId('prev-btn');
    fireEvent.click(prevBtn);
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
  });

  it('should reset the form back to initialValues if clicked on reset button', async () => {
    const { getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    const fillBtn = getByTestId('fill-btn');
    fireEvent.click(fillBtn);
    const resetBtn = getByTestId('reset-btn');
    fireEvent.click(resetBtn);
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
    const loopField = ['trackName', 'trackViewUrl'];
    for (let key of loopField) {
      expect(document.getElementById(key).value).toEqual(String(initialValues[key]));
    }
  });

  it('should show invalid file type message when selected file is not of type image', async () => {
    history.location.pathname = '/add-track/upload';
    const { baseElement, getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );

    const uploadInput = getByTestId('uploader');

    const dummyHTML = new File(['<h1>Welcome</h1>', '<p>World</p>'], 'index.html', { type: 'text/html' });
    fireEvent.change(uploadInput, {
      target: {
        files: [dummyHTML]
      }
    });

    await waitFor(() => expectMessage(baseElement, 'error', INVALID_TYPE_MESSAGE));
  });

  it('should show error popup message when selected file size is greater than 2MB', async () => {
    history.location.pathname = '/add-track/upload';
    const { baseElement, getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );

    const uploadInput = getByTestId('uploader');
    const dummyImage = new File(['somestuff'], 'camera.jpeg', { type: 'image/jpeg' });
    Object.defineProperty(dummyImage, 'size', { value: 2.1 * 1024 * 1024 });
    fireEvent.change(uploadInput, {
      target: {
        files: [dummyImage]
      }
    });

    await waitFor(() => expectMessage(baseElement, 'error', ISLT2M_MESSAGE));
  });

  it('should show message for failed image upload', async () => {
    history.location.pathname = '/add-track/upload';
    const { baseElement, getByTestId } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );

    const uploadInput = getByTestId('uploader');
    const dummyImage = new File(['somestuff2'], 'camera.jpeg', { type: 'image/jpeg' });
    Object.defineProperty(dummyImage, 'size', { value: 10 });
    fireEvent.change(uploadInput, {
      target: {
        files: [dummyImage]
      }
    });

    await waitFor(() => expectMessage(baseElement, 'error', `${dummyImage.name} file upload failed.`));
  });

  // it('should show message for successful image upload', async () => {
  //   history.location.pathname = '/add-track/upload';
  //   const { baseElement, getByTestId } = renderProvider(
  //     <TrackFormContainer
  //       formValues={initialValues}
  //       fillForm={fillSpy}
  //       resetForm={resetSpy}
  //       setFormValues={setFormSpy}
  //       maxWidth={800}
  //     />,
  //     routeProps
  //   );

  //   const uploadInput = getByTestId('uploader');
  //   const dummyImage = new File(['somestuff2'], 'camera.jpeg', { type: 'image/jpeg' });
  //   fireEvent.change(uploadInput, {
  //     target: {
  //       files: [dummyImage]
  //     }
  //   });
  //   console.log({ dummyImage });

  //   await waitFor(() => expectMessage(baseElement, 'success', `${dummyImage.name} file uploaded successfully`));
  // });

  it('should submit values to dummyApi and gives success notification and opens modal', async () => {
    history.location.pathname = '/add-track/upload';
    const { getByTestId, baseElement } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    const uploadInput = getByTestId('uploader');
    const dummyImage = new File(['somestuff2'], 'camera.jpeg', { type: 'image/jpeg' });
    fireEvent.change(uploadInput, {
      target: {
        files: [dummyImage]
      }
    });
    await waitFor(() => expectMessage(baseElement, 'error', `${dummyImage.name} file upload failed.`));

    const submitBtn = getByTestId('submit-btn');
    process.env.TEST_MOCK_STATUS_CODE = 200;
    fireEvent.click(submitBtn);
    await waitFor(() => expect(getNotifElm(baseElement)).toBeInTheDocument());
    await waitFor(() => expect(getByTestId('track-modal')).toBeInTheDocument());
    fireEvent.click(getByTestId('close-modal'));
    await waitFor(() => expect(history.location.pathname).toBe('/add-track/track'));
    notification.destroy();
  });

  it('should submit values to dummyApi and gives error notification', async () => {
    history.location.pathname = '/add-track/upload';
    const { getByTestId, baseElement } = renderProvider(
      <TrackFormContainer
        formValues={initialValues}
        fillForm={fillSpy}
        resetForm={resetSpy}
        setFormValues={setFormSpy}
        maxWidth={800}
      />,
      routeProps
    );
    const uploadInput = getByTestId('uploader');
    const dummyImage = new File(['somestuff2'], 'camera.jpeg', { type: 'image/jpeg' });
    fireEvent.change(uploadInput, {
      target: {
        files: [dummyImage]
      }
    });
    await waitFor(() => expectMessage(baseElement, 'error', `${dummyImage.name} file upload failed.`));

    const submitBtn = getByTestId('submit-btn');
    process.env.TEST_MOCK_STATUS_CODE = 400;
    fireEvent.click(submitBtn);
    await waitFor(() => expect(getNotifElm(baseElement)).toBeInTheDocument());
  });

  it('should maptchDispatchToProps should dispatch action of specific type', () => {
    let dispatchSpy = jest.fn();
    const { setFormValues, resetForm, fillForm } = trackFormContainerCreators;
    const props = mapDispatchToProps(dispatchSpy);

    props.setFormValues({ artworkUrl: 'https://image.com' });
    expect(dispatchSpy).toBeCalledWith(setFormValues({ artworkUrl: 'https://image.com' }));
    props.fillForm();
    expect(dispatchSpy).toBeCalledWith(fillForm());
    props.resetForm();
    expect(dispatchSpy).toBeCalledWith(resetForm());
  });
});
