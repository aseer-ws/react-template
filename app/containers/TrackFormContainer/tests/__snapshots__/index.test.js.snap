// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackFormContainer /> container tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="Container-sc-1vnn72p-0 fyEgoE"
    >
      <div
        class="ant-card ant-card-bordered"
      >
        <div
          class="ant-card-body"
        >
          <header
            class="TrackFormContainer__StyledHeader-fni38k-0 gLzCdi"
          >
            <p
              class="T__StyledText-gjlic1-0 dDQWWc TrackFormContainer__StyledHeaderText-fni38k-1 fhBqzx"
              data-testid="t"
            >
              Add track
            </p>
          </header>
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <div
                class="ant-steps ant-steps-horizontal ant-steps-label-horizontal"
                data-testid="step-parent"
                style="margin-bottom: 2rem;"
              >
                <div
                  class="ant-steps-item ant-steps-item-process TrackFormContainer__CenteredStep-fni38k-3 fFkCEW ant-steps-item-active"
                  data-testid="0_step"
                >
                  <div
                    class="ant-steps-item-container"
                  >
                    <div
                      class="ant-steps-item-tail"
                    />
                    <div
                      class="ant-steps-item-icon"
                    >
                      <span
                        class="ant-steps-icon"
                      >
                        1
                      </span>
                    </div>
                    <div
                      class="ant-steps-item-content"
                    >
                      <div
                        class="ant-steps-item-title"
                      >
                        TRACK
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-steps-item ant-steps-item-wait TrackFormContainer__CenteredStep-fni38k-3 fFkCEW"
                  data-testid="1_step"
                >
                  <div
                    class="ant-steps-item-container"
                  >
                    <div
                      class="ant-steps-item-tail"
                    />
                    <div
                      class="ant-steps-item-icon"
                    >
                      <span
                        class="ant-steps-icon"
                      >
                        2
                      </span>
                    </div>
                    <div
                      class="ant-steps-item-content"
                    >
                      <div
                        class="ant-steps-item-title"
                      >
                        COLLECTION
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-steps-item ant-steps-item-wait TrackFormContainer__CenteredStep-fni38k-3 fFkCEW"
                  data-testid="2_step"
                >
                  <div
                    class="ant-steps-item-container"
                  >
                    <div
                      class="ant-steps-item-tail"
                    />
                    <div
                      class="ant-steps-item-icon"
                    >
                      <span
                        class="ant-steps-icon"
                      >
                        3
                      </span>
                    </div>
                    <div
                      class="ant-steps-item-content"
                    >
                      <div
                        class="ant-steps-item-title"
                      >
                        ARTIST
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-steps-item ant-steps-item-wait TrackFormContainer__CenteredStep-fni38k-3 fFkCEW"
                  data-testid="3_step"
                >
                  <div
                    class="ant-steps-item-container"
                  >
                    <div
                      class="ant-steps-item-tail"
                    />
                    <div
                      class="ant-steps-item-icon"
                    >
                      <span
                        class="ant-steps-icon"
                      >
                        4
                      </span>
                    </div>
                    <div
                      class="ant-steps-item-content"
                    >
                      <div
                        class="ant-steps-item-title"
                      >
                        UPLOAD
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <form
                class="ant-form ant-form-horizontal"
                id="track-form"
              >
                <div
                  class="ant-row ant-form-item"
                  data-testid="trackName_test"
                >
                  <div
                    class="ant-col ant-col-7 ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="trackName"
                      title="Track Name"
                    >
                      Track Name
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-17 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <input
                          class="ant-input"
                          data-testid="trackName"
                          id="trackName"
                          placeholder="Fill in track name"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-form-item"
                  data-testid="trackPrice_test"
                >
                  <div
                    class="ant-col ant-col-7 ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="trackPrice"
                      title="Track Price"
                    >
                      Track Price
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-17 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-input-number"
                        >
                          <div
                            class="ant-input-number-handler-wrap"
                          >
                            <span
                              aria-disabled="false"
                              aria-label="Increase Value"
                              class="ant-input-number-handler ant-input-number-handler-up"
                              role="button"
                              unselectable="on"
                            >
                              <span
                                aria-label="up"
                                class="anticon anticon-up ant-input-number-handler-up-inner"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span
                              aria-disabled="false"
                              aria-label="Decrease Value"
                              class="ant-input-number-handler ant-input-number-handler-down"
                              role="button"
                              unselectable="on"
                            >
                              <span
                                aria-label="down"
                                class="anticon anticon-down ant-input-number-handler-down-inner"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                          <div
                            class="ant-input-number-input-wrap"
                          >
                            <input
                              aria-valuemin="0"
                              aria-valuenow="1"
                              autocomplete="off"
                              class="ant-input-number-input"
                              data-testid="trackPrice"
                              id="trackPrice"
                              placeholder="Fill in track price"
                              role="spinbutton"
                              step="1"
                              value="1"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-form-item"
                  data-testid="primaryGenreName_test"
                >
                  <div
                    class="ant-col ant-col-7 ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="primaryGenreName"
                      title="Track Genre"
                    >
                      Track Genre
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-17 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-select ant-select-single ant-select-show-arrow"
                          data-testid="for"
                          orientation="row"
                        >
                          <div
                            class="ant-select-selector"
                          >
                            <span
                              class="ant-select-selection-search"
                            >
                              <input
                                aria-activedescendant="primaryGenreName_list_0"
                                aria-autocomplete="list"
                                aria-controls="primaryGenreName_list"
                                aria-haspopup="listbox"
                                aria-owns="primaryGenreName_list"
                                autocomplete="off"
                                class="ant-select-selection-search-input"
                                id="primaryGenreName"
                                readonly=""
                                role="combobox"
                                style="opacity: 0;"
                                type="search"
                                unselectable="on"
                                value=""
                              />
                            </span>
                            <span
                              class="ant-select-selection-item"
                              title="Jaz"
                            >
                              Jaz
                            </span>
                          </div>
                          <span
                            aria-hidden="true"
                            class="ant-select-arrow"
                            style="user-select: none;"
                            unselectable="on"
                          >
                            <span
                              aria-label="down"
                              class="anticon anticon-down ant-select-suffix"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                />
                              </svg>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-form-item"
                  data-testid="releaseDate_test"
                >
                  <div
                    class="ant-col ant-col-7 ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="releaseDate"
                      title="Release Date"
                    >
                      Release Date
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-17 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-picker"
                        >
                          <div
                            class="ant-picker-input"
                          >
                            <input
                              autocomplete="off"
                              data-testid="releaseDate"
                              id="releaseDate"
                              placeholder="Select date"
                              readonly=""
                              size="12"
                              title="2022-01-20"
                              value="2022-01-20"
                            />
                            <span
                              class="ant-picker-suffix"
                            >
                              <span
                                aria-label="calendar"
                                class="anticon anticon-calendar"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="calendar"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span
                              class="ant-picker-clear"
                              role="button"
                            >
                              <span
                                aria-label="close-circle"
                                class="anticon anticon-close-circle"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="close-circle"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-row ant-form-item"
                  data-testid="trackViewUrl_test"
                >
                  <div
                    class="ant-col ant-col-7 ant-form-item-label"
                  >
                    <label
                      class="ant-form-item-required"
                      for="trackViewUrl"
                      title="Track URL"
                    >
                      Track URL
                    </label>
                  </div>
                  <div
                    class="ant-col ant-col-17 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <input
                          class="ant-input"
                          data-testid="trackViewUrl"
                          id="trackViewUrl"
                          placeholder="https://track.com"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center TrackFormContainer__SpreadSpace-fni38k-2 hykIwf"
                  style="text-align: center; width: 100%; margin: 1rem auto;"
                >
                  <div
                    class="ant-space-item"
                  />
                </div>
                <div
                  class="ant-divider ant-divider-horizontal"
                  role="separator"
                />
                <div
                  class="ant-row ant-form-item"
                >
                  <div
                    class="ant-col ant-col-24 ant-form-item-control"
                  >
                    <div
                      class="ant-form-item-control-input"
                    >
                      <div
                        class="ant-form-item-control-input-content"
                      >
                        <div
                          class="ant-space ant-space-vertical TrackFormContainer__SpreadSpace-fni38k-2 hykIwf"
                          style="text-align: center; width: 100%;"
                        >
                          <div
                            class="ant-space-item"
                            style="margin-bottom: 10px;"
                          >
                            <div
                              class="ant-space ant-space-horizontal ant-space-align-center"
                            >
                              <div
                                class="ant-space-item"
                                style="margin-right: 8px;"
                              >
                                <button
                                  class="ant-btn"
                                  data-testid="prev-btn"
                                  disabled=""
                                  type="button"
                                >
                                  <span>
                                    Prev
                                  </span>
                                </button>
                              </div>
                              <div
                                class="ant-space-item"
                              >
                                <button
                                  class="ant-btn ant-btn-primary"
                                  data-testid="submit-btn"
                                  type="submit"
                                >
                                  <span>
                                    Next
                                  </span>
                                </button>
                              </div>
                            </div>
                          </div>
                          <div
                            class="ant-space-item"
                          >
                            <div
                              class="ant-space ant-space-horizontal ant-space-align-center"
                            >
                              <div
                                class="ant-space-item"
                                style="margin-right: 8px;"
                              >
                                <button
                                  class="ant-btn"
                                  data-testid="reset-btn"
                                  type="button"
                                >
                                  <span>
                                    Reset
                                  </span>
                                </button>
                              </div>
                              <div
                                class="ant-space-item"
                              >
                                <button
                                  class="ant-btn ant-btn-dashed"
                                  data-testid="fill-btn"
                                  type="button"
                                >
                                  <span>
                                    Fill form
                                  </span>
                                </button>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
